### ISSUE TITLE
No in-place version of where()

### ISSUE THREAD

**Issue by bdusell** [pytorch/pytorch#28329]
## 🚀 Feature
There is no way to perform the functionality of `torch.where()` in-place. This feature would either add a `.where_()` method to `Tensor` or an `out` parameter to the existing `torch.where()`, or both.

## Motivation

All of the usual reasons for doing operations in-place.

## Pitch

As stated above, add a `.where_()` method to `Tensor` or an `out` parameter to the existing `torch.where()`.

## Alternatives

One could use the out-of-place `torch.where()` for a less efficient alternative.

Note that the `.masked_scatter_()` method does not have the same functionality as an in-place `where()`.

**Comment by pietern** [comment]
The in place version would choose the src tensor if the condition is true and take an additional tensor of the same size as argument to choose from if the condition is false, I assume?

This seems like a reasonable additional.

**Comment by bdusell** [comment]
Correct, with the usual broadcasting semantics for the condition and second tensor. The function is basically a no-op for elements where the condition is true and a copy where the condition is false.

**Comment by pietern** [comment]
Are you interested in submitting a PR for this?

**Comment by bdusell** [comment]
Yes.

What is the criterion for adding an underscore method to `Tensor`? There are some point-wise operations that do not have one yet also have a function with an `out` argument. A good example is max: there is no `.max_` method on `Tensor`, but there is a `torch.max` function with an `out` argument.

**Comment by pietern** [comment]
@gchanan Can you comment on this?

**Comment by albanD** [comment]
From what I see:
- inplace operation (`*_`) are only possible if the output is the same size as the input.
- `out=` construction is always possible as the Tensor given to out will be resized to the need of the function (potentially being backed by new memory after the resize).

So if you want an inplace version, you want a `*_` method. Note that the output must be the same size as the input because it's the same Tensor.
The `out=` construction does not guarantee that the result will be written inplace in out. Only that this Tensor will contain the result after the op.

Also `max` is a bit special because even though we could have an inplace version of `torch.max(t1, t2)`, we can't have an inplace version of `torch.max(t1, dim=0)` because the output is not the same size as the input.

**Comment by bdusell** [comment]
I see, so in this case it would make sense to add both `Tensor.where_()` and `torch.where(..., out)`.

**Comment by albanD** [comment]
If we had only one, `Tensor.where_()` feels more useful.

I think there aren't that many people that make use of the `out=` api anymore (mainly because it does not support autograd). And with the good allocators we have now (especially on gpu), reusing buffers is not as useful as it used to be.

**Comment by gchanan** [comment]
I agree with @albanD's take.

Also the `out=` API is broken, see e.g. https://github.com/pytorch/pytorch/issues/8989, so we are generally hesitant to add to it unless there is a good reason (and since inplace covers your usecase, there isn't really a good reason here).

**Comment by LukeLIN-web** [comment]
I also need this function. How I can use? `Tensor.where_()` is working?

**Comment by LukeLIN-web** [comment]
ChatGpt told me I can use 
```
        x.masked_fill_(mask, 0)  # set the values of cached nodes in x to 0
        x += emb  
```

**Comment by albanD** [comment]
Note that you also need to mask your `emb` for this to work!

**Comment by LukeLIN-web** [comment]
> Note that you also need to mask your `emb` for this to work!

I met a problem.
When I used these codes, it shows 
```python
        x.masked_fill_(mask, 0)  # set the values of cached nodes in x to 0
        x += emb  # add the embeddings of the cached nodes to x
        return x
```
RuntimeError: one of the variables needed for gradient computation has been modified by an in-place operation: [torch.cuda.FloatTensor [3, 4]], which is output 0 of ReluBackward0, is at version 2
I change to 
```python
+        out = x.clone()
+        out.masked_fill_(mask, 0)  # set the values of cached nodes in x to 0
+        out += emb  # add the embeddings of the cached nodes to x
+        return out
```
But it cannot keep the gradient.

**Comment by albanD** [comment]
Hi,
The code proposed in https://github.com/pytorch/pytorch/issues/28329#issuecomment-1495525647 is not correct. Could you update your comment above to mention that to make sure people don't get confused by it?

Here is an inplace version that you can use that will work:
```python
a = torch.ones(2, 2, 2)
mask = torch.randn(2, 2, 2).gt(0)
b = torch.zeros(2, 2, 2)

ref_out = torch.where(mask, a, b)

a.masked_scatter_(mask.logical_not(), b)

print((ref_out - a).abs().max())
print(ref_out)
print(a)

```

**Comment by LukeLIN-web** [comment]
> Hi, The code proposed in [#28329 (comment)](https://github.com/pytorch/pytorch/issues/28329#issuecomment-1495525647) is not correct. Could you update your comment above to mention that to make sure people don't get confused by it?
> 
> Here is an inplace version that you can use that will work:
> 
> ```python
> a = torch.ones(2, 2, 2)
> mask = torch.randn(2, 2, 2).gt(0)
> b = torch.zeros(2, 2, 2)
> 
> ref_out = torch.where(mask, a, b)
> 
> a.masked_scatter_(mask.logical_not(), b)
> 
> print((ref_out - a).abs().max())
> print(ref_out)
> print(a)
> ```

Thank you for you help! I delete it.

**Comment by JulianKu** [comment]
> Hi, The code proposed in [#28329 (comment)](https://github.com/pytorch/pytorch/issues/28329#issuecomment-1495525647) is not correct. Could you update your comment above to mention that to make sure people don't get confused by it?
> 
> Here is an inplace version that you can use that will work:
> 
> ```python
> a = torch.ones(2, 2, 2)
> mask = torch.randn(2, 2, 2).gt(0)
> b = torch.zeros(2, 2, 2)
> 
> ref_out = torch.where(mask, a, b)
> 
> a.masked_scatter_(mask.logical_not(), b)
> 
> print((ref_out - a).abs().max())
> print(ref_out)
> print(a)
> ```

In my case, I get wrong results, if I do not mask `b` with `mask.logical_not()` either.
It works if I compute
```
not_mask = mask.logical_not()
a.masked_scatter_(not_mask, b[not_mask])
```
